### PR TITLE
Enhance pickle module and protocols

### LINKED ISSUE #3402: OverflowError: serializing a string larger than 4 GiB requires pickle protocol 4 or higher
Thanks for your good work! MONAI helps me a lot.
When I use PersistentDataset, an error occurs. It seems like error comes from pickle:

Traceback (most recent call last):
  File "/data/zyd/beifen_xinsuoqu_/MedSeg/MedSeg/train/Task900TrainerDP.py", line 296, in <module>
    trainer.train(max_epochs=max_epochs, device=device)
  File "/data/zyd/beifen_xinsuoqu_/MedSeg/MedSeg/train/Task900TrainerDP.py", line 88, in train
    for batch_data in self.train_loader:
  File "/home/zyd/anaconda3/envs/med/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 521, in __next__
    data = self._next_data()
  File "/home/zyd/anaconda3/envs/med/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 561, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
  File "/home/zyd/anaconda3/envs/med/lib/python3.9/site-packages/torch/utils/data/_utils/fetch.py", line 49, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/zyd/anaconda3/envs/med/lib/python3.9/site-packages/torch/utils/data/_utils/fetch.py", line 49, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/zyd/MONAI/monai/data/dataset.py", line 96, in __getitem__
    return self._transform(index)
  File "/home/zyd/MONAI/monai/data/dataset.py", line 288, in _transform
    pre_random_item = self._cachecheck(self.data[index])
  File "/home/zyd/MONAI/monai/data/dataset.py", line 277, in _cachecheck
    torch.save(_item_transformed, temp_hash_file)
  File "/home/zyd/anaconda3/envs/med/lib/python3.9/site-packages/torch/serialization.py", line 379, in save
    _save(obj, opened_zipfile, pickle_module, pickle_protocol)
  File "/home/zyd/anaconda3/envs/med/lib/python3.9/site-packages/torch/serialization.py", line 484, in _save
    pickler.dump(obj)
OverflowError: serializing a string larger than 4 GiB requires pickle protocol 4 or higher

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #3402 .

Followup PR to update below items:
1. Add pickle args to `LRFinder`.
2. Fixed the wrong default protocol to align with `torch.save`, which is `2`. While the default value of pickle is 4:
https://docs.python.org/3/library/pickle.html#pickle.DEFAULT_PROTOCOL
https://pytorch.org/docs/stable/_modules/torch/serialization.html#save
3. Added more description for the string arg for multi-processing.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/optimizers/lr_finder.py [hunk 1]
```diff
@@ -221,7 +231,9 @@ def __init__(
         # Save the original state of the model and optimizer so they can be restored if
         # needed
         self.model_device = next(self.model.parameters()).device
-        self.state_cacher = StateCacher(memory_cache, cache_dir=cache_dir)
```
Comment (wyli) [inline monai/optimizers/lr_finder.py:224]: I feel this is not related to the issue, it's for the model and optimizer, do we want to use other protocols?

Comment (Nic-Ma) [reply monai/optimizers/lr_finder.py:224]: Hi @wyli ,

Thanks for your review and comments.
I think the pickle protocol would be useful if the `self.model.state_dict()` is bigger than 4GB?

Thanks.

**Review by wyli — APPROVED** [review]
Thanks, it looks good to me.
